src/meta/src/hummock/compaction/tier_compaction_picker.rs

```diff
@@ -12,202 +12,390 @@
+                select_level_inputs.push(InputLevel {
+                    level_idx: 0,
+                    level_type: other.level_type,
+                    table_infos: other.table_infos.clone(),
+                });
             }
 
-            if select_level_inputs.len() < self.config.level0_tier_compact_file_number as usize {
-                idx = next_offset;
+            if select_level_inputs
+                .iter()
+                .map(|level| level.table_infos.len())
+                .sum::<usize>()
+                < self.config.level0_tier_compact_file_number as usize
+            {
                 continue;
             }
 
-            level_handlers[0].add_pending_task(self.compact_task_id, &select_level_inputs);
+            if select_level_inputs.len() <= self.config.level0_tier_compact_file_number as usize / 2
+            {
+                continue;
+            }
+
+            for input_level in &select_level_inputs {
+                level_handler.add_pending_task(self.compact_task_id, 0, &input_level.table_infos);
+            }
 
             return Some(CompactionInput {
-                input_levels: vec![InputLevel {
-                    level_idx: 0,
-                    level_type: LevelType::Overlapping as i32,
-                    table_infos: select_level_inputs,
-                }],
+                input_levels: select_level_inputs,
                 target_level: 0,
-                target_sub_level_id: 0,
+                target_sub_level_id: level.sub_level_id,
             });
         }
         None
     }
-}
 
-pub struct LevelCompactionPicker {
-    compact_task_id: u64,
-    target_level: usize,
-    overlap_strategy: Arc<dyn OverlapStrategy>,
-    config: Arc<CompactionConfig>,
-}
+    fn pick_trivial_move_file(
+        &self,
+        l0: &OverlappingLevel,
+        level_handlers: &mut [LevelHandler],
+    ) -> Option<CompactionInput> {
+        for (idx, level) in l0.sub_levels.iter().enumerate() {
+            if level.level_type == LevelType::Overlapping as i32 || idx + 1 >= l0.sub_levels.len() {
+                continue;
+            }
 
-#[derive(Default)]
-pub struct TargetFilesInfo {
-    tables: Vec<SstableInfo>,
-    table_ids: HashSet<u64>,
-    compaction_bytes: u64,
-}
+            if l0.sub_levels[idx + 1].level_type == LevelType::Overlapping as i32 {
+                continue;
+            }
 
-impl TargetFilesInfo {
-    fn add_tables(&mut self, new_add_tables: Vec<SstableInfo>) {
-        for table in new_add_tables {
-            if self.table_ids.contains(&table.id) {
+            let min_overlap_picker = MinOverlappingPicker::new(
+                self.compact_task_id,
+                0,
+                0,
+                self.overlap_strategy.clone(),
+            );
+
+            let (select_tables, target_tables) = min_overlap_picker.pick_tables(
+                &l0.sub_levels[idx + 1].table_infos,
+                &level.table_infos,
```

Comment:
Shall we swap these two parameters? In other words, use `&level.table_infos` as select_tables and `&l0.sub_levels[idx + 1].table_infos` as taget_tables.

Reply:
No. Because we must use higher level to overlap more bottom level.